pkg/cliplugins/workspace/plugin/kubeconfig.go

```diff
@@ -82,468 +105,215 @@ func NewKubeConfig(opts *Options) (*KubeConfig, error) {
-		Cluster:  kcpPreviousWorkspaceContextKey,
-		AuthInfo: currentContextAuthName,
+	if err := kc.modifyConfig(newKubeConfig); err != nil {
+		return err
 	}
 
-	return clientcmd.ModifyConfig(kc.configAccess, *kc.startingConfig, true)
+	return kc.currentWorkspace(ctx, newServerHost, workspaceType)
 }
 
 // CurrentWorkspace outputs the current workspace.
-func (kc *KubeConfig) CurrentWorkspace(ctx context.Context, opts *Options) error {
-	org, workspaceName, err := kc.getCurrentWorkspace(opts)
+func (kc *KubeConfig) CurrentWorkspace(ctx context.Context) error {
+	config, err := clientcmd.NewDefaultClientConfig(*kc.startingConfig, kc.overrides).ClientConfig()
 	if err != nil {
 		return err
 	}
 
-	workspacePrettyName := workspaceName
-	if !org.Empty() {
-		workspaceDirectoryRestConfig, err := kc.workspaceDirectoryRestConfigFromCurrentContext(opts, true)
-		if err != nil {
-			return err
-		}
-
-		tenancyClient, err := tenancyclient.NewForConfig(workspaceDirectoryRestConfig)
-		if err != nil {
-			return err
-		}
-
-		if ws, err := getWorkspaceFromInternalName(ctx, workspaceName, tenancyClient); err != nil {
-			_ = outputCurrentWorkspaceMessage(org, workspacePrettyName, workspaceName, opts)
-			return err
-		} else {
-			if opts.Scope == registry.PersonalScope {
-				workspacePrettyName = ws.Name
-			}
-		}
-	}
-
-	return outputCurrentWorkspaceMessage(org, workspacePrettyName, workspaceName, opts)
+	return kc.currentWorkspace(ctx, config.Host, "")
 }
 
-// ListWorkspaces outputs the list of workspaces of the current user
-// (kubeconfig user possibly overridden by CLI options).
-func (kc *KubeConfig) ListWorkspaces(ctx context.Context, opts *Options) error {
-	workspaceDirectoryRestConfig, err := kc.workspaceDirectoryRestConfigFromCurrentContext(opts, false)
-	if err != nil {
-		return err
-	}
-
-	tenancyClient, err := tenancyclient.NewForConfig(workspaceDirectoryRestConfig)
+func (kc *KubeConfig) currentWorkspace(ctx context.Context, host, workspaceType string) error {
+	_, clusterName, err := parseClusterURL(host)
 	if err != nil {
+		_, err = fmt.Fprintf(kc.Out, "Current workspace is the URL %q\n", host)
 		return err
 	}
 
-	result := tenancyClient.TenancyV1beta1().RESTClient().Get().Resource("workspaces").SetHeader("Accept", strings.Join([]string{
-		fmt.Sprintf("application/json;as=Table;v=%s;g=%s", metav1.SchemeGroupVersion.Version, metav1.GroupName),
-		fmt.Sprintf("application/json;as=Table;v=%s;g=%s", metav1beta1.SchemeGroupVersion.Version, metav1beta1.GroupName),
-		"application/json",
-	}, ",")).Do(ctx)
-
-	var statusCode int
-	if result.StatusCode(&statusCode).Error() != nil {
-		return result.Error()
-	}
-
-	if statusCode != http.StatusOK {
-		rawResult, err := result.Raw()
-		if err != nil {
-			return err
+	parentClusterName, workspaceName := clusterName.Split()
+	workspacePrettyName := workspaceName
+	if !parentClusterName.Empty() && clusterName.HasPrefix(tenancyv1alpha1.RootCluster) {
```

Comment:
Not sure I understand the meaning of this condition:
``` golang
&& clusterName.HasPrefix(tenancyv1alpha1.RootCluster)
```

Reply:
To only deal with root:.... and not system:...?

Reply:
```
	parentClusterName, workspaceName := clusterName.Split()
	workspacePrettyName := workspaceName
	if clusterName.HasPrefix(tenancyv1alpha1.RootCluster) && clusterName != tenancyv1alpha1.RootCluster {
		ws, err := getWorkspaceFromInternalName(ctx, workspaceName, kc.client.Cluster(parentClusterName))
		if err == nil {
			workspacePrettyName = ws.Name
		}
	}
```